### PR TITLE
feat(denommetadata): inject denom metadata to ibc transfers from RA to Hub

### LINKED ISSUE #454: feat(denommetadata) send local Denom Metadata to Hub through IBC transfer

### PR DESCRIPTION
**Pull request by zale144**
# PR Standards

## Opening a pull request should be able to meet the following requirements

---

Closes #454 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: x/hub/keeper/keeper.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	// Check if the packet is destined for a hub
+	chainID, err := k.ExtractChainIDFromChannel(ctx, hubPortOnRollapp, hubChannelOnRollapp)
+	if err != nil {
+		return nil, err
+	}
+	k.SetHub(ctx, types.Hub{
+		Id:               chainID,
+		ChannelId:        hubChannelOnRollapp,
+		RegisteredDenoms: nil,
+	})
```
Comment (danwt) [inline x/hub/keeper/keeper.go:49]: not really correct because how do you know this is the hub?
and why are you resetting denoms to nil?

File: x/hub/keeper/keeper.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	if hub.ChannelId == "" {
+		return nil, errorsmod.Wrapf(types.ErrGenesisEventNotTriggered, "empty channel id: hub id: %s", chainID)
+	}
+	// check if the channelID matches the hubID's channelID
+	if hub.ChannelId != hubChannelOnRollapp {
+		return nil, errorsmod.Wrapf(
+			types.ErrMismatchedChannelID,
+			"channel id mismatch: expect: %s: got: %s", hub.ChannelId, hubChannelOnRollapp,
+		)
+	}
```
Comment (danwt) [inline x/hub/keeper/keeper.go:65]: there is no such thing as genesis event anymore

**Review by danwt — CHANGES_REQUESTED** [review]
Makes sense I think just a couple things

Not sure about active/inactive as names. Can you think of alternatives? And they need to be documented exactly what each one means.

Also, with errors. Please prefer https://pkg.go.dev/github.com/dymensionxyz/gerr-cosmos@v0.1.2/gerrc#pkg-variables now. For example, you can wrap the AlreadyExists one there, instead of defining new ones

Also, please don't call the types. 'middleware' because they don't implement porrtypes middleware. Call them IBCModule or ICS4Wrapper

Final point: please don't implement some kind of canonical channel right now. You should write your code to assume that this is handled by some other component, and we will make an issue to fix that properly

File: x/hub/keeper/keeper.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+func (k Keeper) GetHub(ctx sdk.Context, id string) (hub types.Hub, found bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get([]byte(id))
+	if bz == nil {
+		return hub, false
+	}
+	k.cdc.MustUnmarshal(bz, &hub)
+	return hub, true
+}
```
Comment (danwt) [inline x/hub/keeper/keeper.go:81]: not sure why hub has a id
There will surely only ever be one

File: x/hub/keeper/keeper.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+		return "", fmt.Errorf("extract clientID from channel: %w", err)
+	}
+
+	tmClientState, ok := clientState.(*ibctypes.ClientState)
+	if !ok {
+		return "", nil
+	}
+
+	return tmClientState.ChainId, nil
+}
```
Comment (danwt) [inline x/hub/keeper/keeper.go:66]: please favour our shared utils package now for stuff like this
https://github.com/dymensionxyz/sdk-utils/blob/68179fdc62bb54adb05654ee6b9d1ede4bc1b672/utils/uibc/funcs.go#L28-L49
It's already used in both the hub and rdk on main branches

File: proto/hub/hub.proto [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+option go_package = "github.com/dymensionxyz/dymension-rdk/x/hub/types";
+
+// Hub is a proto message that represents the metadata of the Hub
+message Hub {
+  // id is the unique identifier of the Hub
+  string id = 1;
+
+  // channel_id is the unique identifier of the channel that the Hub is connected to
+  string channel_id = 2;
+
```
Comment (danwt) [inline proto/hub/hub.proto:15]: not needed right

**Review by danwt — CHANGES_REQUESTED** [review]
I'm not sure I'm on board with this architecturally.

I think the denom meta data module should just assume there is one hub and keep track of the registered denoms for it by itself

No need for hub keeper

**Comment by zale144** [comment]
> I'm not sure I'm on board with this architecturally.
> 
> I think the denom meta data module should just assume there is one hub and keep track of the registered denoms for it by itself
> 
> No need for hub keeper

`keep track of the registered denoms for it by itself` 
so how do we prevent the rollapp from sending the denom metadata then?

I think I already addressed a similar comment you posted: even though the hub can always be the same, technically it is still possible the hub chain can do a fork, which I believe would make it's chain ID change.

See: https://github.com/dymensionxyz/dymension-rdk/issues/465

**Comment by danwt** [comment]
>so how do we prevent the rollapp from sending the denom metadata then?
WDYM?
I'm just saying to store what you currently store on the hub module in the denom meta module

Re https://github.com/dymensionxyz/dymension-rdk/issues/465 ok maybe. IDK tbh I would have to research. Seems like a delicate thing. Can you just write this PR so that it assumes that problem is solved by another component. Otherwise you're solving too many problems at once and stuff is likely to get reworked/changed anyways

Basically just write your code to have minimal assumptions so that the rest of the codebase can evolve around it independently. Here you're placing quite a lot of assumptions on the management of the hub wrt. forks etc, which is out of scope.

**Review by danwt — APPROVED** [review]
great work!

**Review by mtsitrin — APPROVED** [review]
